vim25/xml/util/util.go

```diff
@@ -0,0 +1,241 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package util implements some helper functions for marshaling/unmarshaling
+// vim objects to/from XML.
+package util
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"reflect"
+	"strings"
+
+	"github.com/vmware/govmomi/vim25"
+	"github.com/vmware/govmomi/vim25/types"
+	"github.com/vmware/govmomi/vim25/xml"
+)
+
+// VimObjectsFromBytes returns all vim objects decoded from the provided bytes.
+//
+// This is a helper function that wraps VimObjectsFromDecoder. Please see
+// the documentation on VimObjectsFromDecoder for additional information.
+func VimObjectsFromBytes(v []byte) (<-chan interface{}, <-chan error) {
+	return VimObjectsFromReader(bytes.NewReader(v))
+}
+
+// VimObjectsFromString returns all vim objects decoded from the provided
+// string.
+//
+// This is a helper function that wraps VimObjectsFromDecoder. Please see
+// the documentation on VimObjectsFromDecoder for additional information.
+func VimObjectsFromString(v string) (<-chan interface{}, <-chan error) {
+	return VimObjectsFromReader(strings.NewReader(v))
+}
+
+// VimObjectsFromReader returns all vim objects decoded from the provided input
+// stream.
+//
+// This is a helper function that wraps VimObjectsFromDecoder. Please see
+// the documentation on VimObjectsFromDecoder for additional information.
+func VimObjectsFromReader(r io.Reader) (<-chan interface{}, <-chan error) {
+	return VimObjectsFromDecoder(xml.NewDecoder(r))
+}
+
+// VimObjectsFromDecoder returns all vim objects from the provided decoder.
+//
+// Any decode errors are sent on the error channel.
+//
+// Both channels are closed when io.EOF is reached.
+//
+// The decoder's TypeFunc is used to construct types from their XML data.
+// If the decoder's TypeFunc is nil then vim25.TypeFunc is used.
+func VimObjectsFromDecoder(d *xml.Decoder) (<-chan interface{}, <-chan error) {
+
+	var (
+		chanObj = make(chan interface{})
+		chanErr = make(chan error)
+	)
+
+	// If the decoder did not provide a TypeFunc then use the one from the
+	// vim25.types package.
+	if d.TypeFunc == nil {
+		d.TypeFunc = types.TypeFunc()
+	}
+
+	// Returns the value of the "xsi:type" attribute if it exists, otherwise
+	// the value of the "type" attribute, otherwise an empty string.
+	getVimTypeName := func(attrs []xml.Attr) string {
+		var vimTypeName string
+		for _, a := range attrs {
+			if a.Name.Local == "type" {
+				if a.Name.Space == xml.SchemaInstanceURI {
+					vimTypeName = a.Value
+				} else if vimTypeName == "" {
+					vimTypeName = a.Value
+				}
+			}
+		}
+		return vimTypeName
+	}
+
+	// Decode the input stream in the background, returning tokens/errors on
+	// the chanObj and chanErr channels.
+	go func() {
+
+		// Keep iterating until there are no more tokens.
+		for {
+
```

Comment:
nit: remove newline (same for other places)

Reply:
I prefer the newlines for readability. Unless you feel strongly, I prefer to leave them.

Reply:
I don't feel strongly about it, but I personally do not add newlines after a line ending w/ a brace as I find the indent provides readability. Either way is fine with me though.